Dev 4.x newbanpolicy

sidhujag opened the pull request:
new ban policy to avoid pings all the time but to only have to ping when the masternode is in the winners list up to 20 in the past and 10 in the future (so 30 masternodes effectively maximum that are pinging at any given time).

1) lastPing would set everytime a ping is recognized but we instead ensure that pings are only accepted if the masternode is in the winners list (20 in past and 10 in future).
2) If master node is checked (every 10 seconds) and the masternode isn't in the winners list criteria above, then we set lastPing to empty and also skip checks that depend on lastPing  as they will only be valid once a ping exists. Never expire masternodes until they are actually being checked and are in the masternode winners list.
3) If a masternode is expired, likely they would be removed from future payments, they will get paid for this round.
4) changed the expiry params, minimum time between pings is 60 seconds and masternodes have 10 minutes to ping the network before expiry.
5) If masternode expires, allow up to 2 expiries, after 3 expiries you will goto expired mode, if ping works next round masternode should revive again.
6) If masternode expired 5 or more times, it will be put in new start mode meaning it is removed and required to start again from scratch.
7) Only start counting the expiry once 6 signatures attest to the winners list to ensure that the masternode is in every nodes winners list likely

laanwj committed:
wip on new ban policy

sidhujag committed:
add pingretries to cmasternode

sidhujag committed:
start masternode preenabled and switch to enabled on ping check from broadcast

willyko commented:
What's the scale of the testnet needed to verify that this is working as intended?  Also, is it possible for the 30 masternodes on the current winner list to collude? 30 masternodes seem achievable and as you reject more nodes, your chance of getting a node into the list is higher as well

sidhujag committed:
display retries

sidhujag committed:
reset ping retries once successful ping and also ensure mn is in winnerslist

if mn is in winnerlist + 6 signatures a ping is accepted, if mn is in winners list between 0 and 6 signatures it wont be valid but it will not DOS ban peer, if  you are outside of the list entirely you should ban the peer.

The idea is that once you get into the winners list and have enough signatures its ok to start pinging. This avoids the race condition of pingers being banned and being accepted because thre is a delay between having 0 and 6 signatures, outside of the winners list

sschepis commented:
While the code looks solid to me, it's hard to make a clear determination without comparing against unit tests. Can we get some unit tests written for this code? I also suggest we make it a policy not to check in new code/functionality into syscoin core that doesn't come with unit tests. Thoughts?

sidhujag commented:
> What's the scale of the testnet needed to verify that this is working as intended? Also, is it possible for the 30 masternodes on the current winner list to collude? 30 masternodes seem achievable and as you reject more nodes, your chance of getting a node into the list is higher as well

The list is deterministically chosen, im not sure how you would collude or what the purpose of colluding is? The list is chosen in the same way we didn't modify that.

sidhujag committed:
change to 2 to test

sidhujag committed:
remove unused vars

sidhujag committed:
accept first ping on new broadcast, dont check winners list for that

sidhujag committed:
dont expire if lastPing is empty

sidhujag committed:
move enabled after checks and fix fSentinelIsCurrent

sidhujag committed:
move force check after enabled set

sidhujag committed:
reset retries only if mn enabled

sidhujag committed:
remove ping expired/expired states, either your enabled or not

sidhujag committed:
remove calls to isexpired and issentinelpingexpired

sidhujag committed:
idb check removed from miner

sidhujag committed:
rmove status print

sidhujag committed:
remove sentinel ping activation checks

sidhujag committed:
comment code out TODO put it back after test

sidhujag committed:
dont send ping unless in winners list with 6 sigs

sidhujag committed:
comment out ping

sidhujag committed:
enabled by default until theres a ping

sidhujag committed:
add sentinel ping expired state back

sidhujag committed:
refactor

sidhujag committed:
move preenabled check above

sidhujag committed:
relay if sentinel expired

sidhujag committed:
brackets around logic

sidhujag committed:
remove dbg

sidhujag committed:
ping from sentinelping for test

sidhujag committed:
some dbg

sidhujag committed:
allow pinging during preenabled state

sidhujag committed:
Revert "some dbg"

This reverts commit e6b06fe0ced0270297f6cce812825f76c5565bbf.

sidhujag committed:
dont increment npingretries if ourmn

sidhujag committed:
fix sentinelping call, call check every 10 min

sidhujag committed:
fix common lib

sidhujag committed:
move corewrite to server

sidhujag committed:
run mnchecks before calls that use the information

sidhujag committed:
is mn is not synced set to preenabled

sidhujag committed:
ensure we lock main first in istransactionvalid

sidhujag committed:
move pingretries to masternode info

sidhujag committed:
set pingretries to mnb list

sidhujag committed:
store ping retries in setmnping

sidhujag committed:
check retries from broadcast

sidhujag committed:
use references not copy

sidhujag committed:
some helpful debug

sidhujag committed:
dbg

sidhujag committed:
remove dbg

sidhujag committed:
allow force check on masternodemanager::Check

pass false by default but pass in true in the rpc calls and istransactionvalid. Don't update npingretries if passed in true, only when its not forced do we update that variable to avoid adding it multiple times

sidhujag committed:
set sentinel valid by default on broadcast

sidhujag committed:
set max retries to 60 and maintenance to every min

sidhujag committed:
only check ping in winners list if winners list is synced

sidhujag committed:
check winnerslist on sending ping only if list is synced

sidhujag committed:
serialize ping retries

sidhujag committed:
manage local ping retries

sidhujag committed:
check entire list is synced not just mn list

sidhujag committed:
startup preenabled for 10s

sidhujag committed:
remove dbg

sidhujag committed:
send ping only if not preenabled and is synced

sidhujag committed:
only check masternodes if they are in the winnerslist (or not enabled)

sidhujag committed:
output for not being in winners list in check

sidhujag committed:
use adjusted time instead of lastping for preenabled check

sidhujag committed:
use function already created IsBroadcastedWithin

sidhujag committed:
future sig time check fix

sidhujag committed:
give atleast MASTERNODE_SENTINEL_PING_MAX_SECONDS/2 seconds to propogate

sidhujag committed:
update to MASTERNODE_MIN_MNB_SECONDS

sidhujag committed:
also adjust broadcast check

sidhujag committed:
flip logic

sidhujag committed:
fix switching from pre to enabled state

sidhujag committed:
const references

sidhujag committed:
remove mn cache's

sidhujag committed:
update proto and hrp

sidhujag committed:
set sigs to 6 again

sidhujag committed:
update get to 1.8.23
